external/storm-kafka-new-consumer-api/src/main/java/org/apache/storm/kafka/spout/KafkaSpout.java

```diff
@@ -0,0 +1,457 @@
+        ackCommitLock = new ReentrantLock();
+        maxRetries = kafkaSpoutConfig.getMaxTupleRetries();
+
+        // Kafka consumer
+        kafkaConsumer = new KafkaConsumer<>(kafkaSpoutConfig.getKafkaProps(),
+                kafkaSpoutConfig.getKeyDeserializer(), kafkaSpoutConfig.getValueDeserializer());
+        kafkaConsumer.subscribe(kafkaSpoutConfig.getSubscribedTopics(), new KafkaSpoutConsumerRebalanceListener());
+
+        // Create commit offsets task
+        if (!kafkaSpoutConfig.isConsumerAutoCommitMode()) {     // If it is auto commit, no need to commit offsets manually
+            createCommitOffsetsTask();
+        }
+    }
+
+    // ======== Commit Offsets Task =======
+
+    private void createCommitOffsetsTask() {
+        commitOffsetsTask = Executors.newSingleThreadScheduledExecutor(commitOffsetsThreadFactory());
+        commitOffsetsTask.scheduleAtFixedRate(new Runnable() {
+            @Override
+            public void run() {
+                commit = true;
+            }
+        }, 1000, kafkaSpoutConfig.getOffsetsCommitFreqMs(), TimeUnit.MILLISECONDS);
+    }
+
+    private ThreadFactory commitOffsetsThreadFactory() {
+        return new ThreadFactory() {
+            @Override
+            public Thread newThread(Runnable r) {
+                return new Thread(r, "kafka-spout-commit-offsets-thread");
+            }
+        };
+    }
+
+    // ======== Next Tuple =======
+
+    @Override
+    public void nextTuple() {
+        if(commit) {
+            commitAckedTuples();
+        } else if (retry()) {              // Don't process new tuples until the failed tuples have all been acked
+            retryFailedTuples();
+        } else {
+            emitTuples(poll());
+        }
+    }
+
+    private ConsumerRecords<K, V> poll() {
+        final ConsumerRecords<K, V> consumerRecords = kafkaConsumer.poll(kafkaSpoutConfig.getPollTimeoutMs());
+        LOG.debug("Polled [{]} records from Kafka", consumerRecords.count());
+        return consumerRecords;
+    }
+
+    private void emitTuples(ConsumerRecords<K, V> consumerRecords) {
+        for (TopicPartition tp : consumerRecords.partitions()) {
+            final Iterable<ConsumerRecord<K, V>> records = consumerRecords.records(tp.topic());     // TODO Decide if want to give flexibility to emmit/poll either per topic or per partition
+            for (ConsumerRecord<K, V> record : records) {
+                final Values tuple = tupleBuilder.buildTuple(record);
+                final org.apache.storm.kafka.spout.MessageId messageId = new org.apache.storm.kafka.spout.MessageId(record);                                  // TODO don't create message for non acking mode. Should we support non acking mode?
+                collector.emit(kafkaSpoutStream.getStreamId(), tuple, messageId);           // emits one tuple per record
+                emittedTuples.put(messageId, tuple);
+                LOG.info("HMCL - Emitted tuple for record {}", record);
```

Comment:
Please remove the "HMCL - " at least and ideally make it as debug, or just delete it all together.